### PR TITLE
 Updated EAD processing (critical and non-critical)

### PR DESCRIPTION
**Pull request by gselander**
Addressing #275

### CODE REVIEW

**Comment by emanjon** [comment]
Issue needed.

File: draft-ietf-lake-edhoc.md [hunk 1]
```diff
@@ -585,7 +585,11 @@ ead = 1* (
 )
 ~~~~~~~~~~~
 
-A security application using external authorization data need to register an ead_label, specify the ead_value format for each message (see {{iana-ead}}), and describe processing and security considerations.
+A security application using external authorization data need to register an ead_label, specify the ead_value formatting for each message (see {{iana-ead}}), and describe processing and security considerations.
+
+An EAD field can be either critical or non-critical, determined by the sign of ead_label: negative means critical; non-negative means non-critical. If an endpoint does not support a received critical EAD field, the EDHOC protocol MUST be discontinued.
```
Comment (emanjon) [inline draft-ietf-lake-edhoc.md:590]: "If an endpoint does not support a received critical EAD field, the EDHOC protocol MUST be discontinued."

This needs more discussion and work. The handling of critical extensions in C.509 is to my understanding confusing and problematic. Both in the  specification and in implementations.

What does support means. Can an endpoint supporting the EAD ignore it.

**Review by emanjon — COMMENTED** [review]
I think critical ead is a very good discussion to have. I would suggest that we have a discussion and add any outcome to -16 shortly after IETF 114.

File: draft-ietf-lake-edhoc.md [hunk 2]
```diff
@@ -585,7 +585,11 @@ ead = 1* (
...
+
+The EAD specification describes if an EAD field is always critical, always non-critical, or if it can be decided by the application in the sending endpoint. In the latter case, one negative and one non-negative instance of ead_label need to be registered.
```
Comment (emanjon) [inline draft-ietf-lake-edhoc.md:592]: The specification likely MUST describe

File: draft-ietf-lake-edhoc.md [hunk 3]
```diff
@@ -1465,7 +1469,7 @@ IANA has created a new registry entitled "EDHOC Error Codes" under the new group
 
 ## EDHOC External Authorization Data Registry {#iana-ead}
 
-IANA has created a new registry entitled "EDHOC External Authorization Data" under the new group name "Ephemeral Diffie-Hellman Over COSE (EDHOC)". The registration procedure is "Specification Required". The columns of the registry are Label, Message, Description, and Reference, where Label is an integer and the other columns are text strings.
+IANA has created a new registry entitled "EDHOC External Authorization Data" under the new group name "Ephemeral Diffie-Hellman Over COSE (EDHOC)". The registration procedure is "Specification Required". The columns of the registry are Label, Description, and Reference, where Label is an integer and the other columns are text strings.
```
Comment (emanjon) [inline draft-ietf-lake-edhoc.md:1472]: This should likely be non-negative integer now.

Comment (emanjon) [reply draft-ietf-lake-edhoc.md]: I did some changes to the registration procedure and the processing regarding critical EAD fields. I think the current version could be included in -15

**Comment by gselander** [comment]
Should we allow or forbid that ead_label  = k and ead_label = -k correspond to different ead-values?

For example, if an (ead_label, ead_value) is always critical: 

* Say, ead_value is an ACE Access Token with ead_label = -4, always critical.

Then we can either:
1. forbid the use of ead_label = 4 altogether, or 
2. allocate an always non-critical EAD to ead_label = 4, say, ead_value is "service indication" as in #275

In case 1 we will have some "holes" in the register. One positive integer blocked for each always critical (ead_label, ead_value) , one negative blocked for each always non-critical (ead_label, ead_value), and 0 cannot be used. (It is not clear to me that there are many always non-critical, but always critical is probably common.)

In case 2, we could still have the registration policy that: if an EAD can be either critical or non-critical then two ead_label with same absolute value should be reserved for it.  

What is the main risk with case 2? If someone misunderstands or by accident changes sign on an always /non-/critical EAD this results most likely in an error due to wrong ead_value (like in the example above), but potentially in an unintended EAD message. (Another registration policy could be that the ead_values corresponding to ead_labels k and -k should have incompatible CDDL so the error is discovered ...) The EAD specification needs anyway to specify what are the processing rules and if you violate those more or less anything could happen. 

Comments?

**Comment by gselander** [comment]
BTW, we may want a term for the ordered pair ead_label, ead_value). The term "EAD" or "EAD field" is referring to the message field, which may consist of multiple (ead_label, ead_value).

**Comment by emanjon** [comment]
> What is the main risk with case 2?

Just seems like overoptimization to me.

**Comment by gselander** [comment]
Wait with this pending design team meeting, see #275.

**Comment by gselander** [comment]
I made a proposal. The text now include the concepts "EAD item" and "critical"/"non-critical" EAD item. There is also more details about how applications makes use of EAD: 
* each application registers its own ead_labels
* ead_label is a positive integer associated to a particular ead_value
* if the application transports different ead_values then multiple ead_labels need to be registered
* the application specifies how EAD items are transported in EDHOC messages, a particular EAD item may be transported in multiple messages (i.e. different EAD_x).

I'll merge this into -15 and look forward to comments!

**Pull request merged**